posix: pthread: fix `zephyr_to_posix_priority` assert test

ycsin opened the pull request:
If `z_prio` is negative and we want to make sure that it is
within `[-CONFIG_NUM_COOP_PRIORITIES, -1]`, we should invert
its sign and make sure that it is `<=`
`CONFIG_NUM_COOP_PRIORITIES` instead of `<`.

Changed the variable type of the priority in the args and the
return type of the conversion functions to `int`, as both
Zephyr's priority & POSIX's `sched_priority` has type `int`.

Made the conversion functions non-static and added ztests for
them to make sure that they work across the full range of
Zephyr priorities.

cfriedt commented:
@ycsin - I haven't checked, but if this was in previous releases, it may be good to preemptively create an issue in case a backport is required. No code changes needed though.

ycsin commented:
> @ycsin - I haven't checked, but if this was in previous releases, it may be good to preemptively create an issue in case a backport is required. No code changes needed though.

The refactor that introduced this change happens after v3.5

ycsin commented on lib/posix/pthread.c:
```diff
@@ -222,15 +222,16 @@ static bool is_posix_policy_prio_valid(uint32_t priority, int policy)
 {
 	if (z_prio < 0) {
-		__ASSERT_NO_MSG(z_prio < CONFIG_NUM_COOP_PRIORITIES);
+		__ASSERT_NO_MSG(-z_prio <= CONFIG_NUM_COOP_PRIORITIES);
```
The change here was
```c
		__ASSERT_NO_MSG(-z_prio < CONFIG_NUM_COOP_PRIORITIES);
```
which make sure that the Zephyr priority (`z_prio`) is within `[-1, -(CONFIG_NUM_COOP_PRIORITIES - 1)]`, but Zephyr COOP priorities ranges from `[-1, -CONFIG_NUM_COOP_PRIORITIES]` instead, so change that to `<=`

ycsin committed:
posix: pthread: fix `zephyr_to_posix_priority` assert test

If `z_prio` is negative and we want to make sure that it is
within `[-CONFIG_NUM_COOP_PRIORITIES, -1]`, we should invert
its sign and make sure that it is `<=`
`CONFIG_NUM_COOP_PRIORITIES` instead of `<`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: pthread: fix typos

- 'piority' should be 'priority'
- COOP should start from `-CONFIG_NUM_COOP_PRIORITIES`

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: pthread: priority should be of `int` type

Changed the variable type of the priority in the args and the
return type of the conversion functions to `int`, as both
Zephyr's priority & POSIX's `sched_priority` has type `int`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
posix: pthread: test the priority conversion functions

Made the conversion functions non-static and added ztests for
them to make sure that they work across the full range of
Zephyr priorities.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented:
I've decided to split the commit into multiple smaller ones so each commit has a smaller scope, the end result is the same. Sorry for the multiple force pushes, can I get the approvals again please? Thanks!

@cfriedt @jukkar @bjarki-trackunit

jukkar commented on lib/posix/pthread.c:
```diff
@@ -242,7 +242,7 @@ static uint32_t zephyr_to_posix_priority(int32_t z_prio, int *policy)
 	if (policy == SCHED_FIFO) {
-		/* COOP: highest [CONFIG_NUM_COOP_PRIORITIES, -1] lowest */
+		/* COOP: highest [-CONFIG_NUM_COOP_PRIORITIES, -1] lowest */
 		__ASSERT_NO_MSG(priority < CONFIG_NUM_COOP_PRIORITIES);
```
Why not `<=` here?

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -222,15 +222,16 @@ static bool is_posix_policy_prio_valid(uint32_t priority, int policy)
 
-static uint32_t zephyr_to_posix_priority(int32_t z_prio, int *policy)
+/* Non-static so that they can be tested in ztest */
+int zephyr_to_posix_priority(int z_prio, int *policy)
```
POSIX priorities are only >= 0.

I guess it would be fine to return a negative value to signal an error. I thought there were assertions to that effect though.

cfriedt commented on lib/posix/pthread.c:
```diff
@@ -697,7 +699,7 @@ int pthread_attr_init(pthread_attr_t *_attr)
 int pthread_getschedparam(pthread_t pthread, int *policy, struct sched_param *param)
 {
-	uint32_t priority;
+	int priority;
```
POSIX priorities are >= 0

cfriedt commented on tests/posix/common/src/pthread.c:
```diff
@@ -215,6 +215,39 @@ void *thread_top_term(void *p1)
 }
 
+/* Test the internal priority conversion functions */
+int zephyr_to_posix_priority(int z_prio, int *policy);
+int posix_to_zephyr_priority(int priority, int policy);
```
POSIX priorities are >= 0

ycsin committed:
posix: pthread: use `is_posix_policy_prio_valid` to check POSIX priority

Use the existing `is_posix_policy_prio_valid()` function to
verify the POSIX's priority in the conversion functions.

Changed the `priority` arg of `is_posix_policy_prio_valid` to
`int` since that is the output of `sched_get_priority_min` &
`sched_get_priority_max`.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on lib/posix/pthread.c:
```diff
@@ -215,39 +215,40 @@ void __z_pthread_cleanup_pop(int execute)
 	*policy = (z_prio < 0) ? SCHED_FIFO : SCHED_RR;
-	return ZEPHYR_TO_POSIX_PRIORITY(z_prio);
+	priority = ZEPHYR_TO_POSIX_PRIORITY(z_prio);
+	__ASSERT_NO_MSG(is_posix_policy_prio_valid(priority, *policy));
```
@cfriedt I've added another commit to check the POSIX priority going out (and coming in, in the other function), should be watertight now?

ycsin commented on lib/posix/pthread.c:
```diff
@@ -215,39 +215,40 @@ void __z_pthread_cleanup_pop(int execute)
 }
 
-static bool is_posix_policy_prio_valid(uint32_t priority, int policy)
+static bool is_posix_policy_prio_valid(int priority, int policy)
```
The `int sched_get_priority_min()` function always returns 0, if `policy` is valid.
